core/common/test/format/DateTimeFormatTest.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2019-2023 JetBrains s.r.o. and contributors.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime.test.format
+
+import kotlinx.datetime.*
+import kotlinx.datetime.format.*
+import kotlin.test.*
+
+class DateTimeFormatTest {
+    @Test
+    fun testPythonDateTime() {
+        val dateTimes = buildMap<LocalDateTime, Pair<String, Set<String>>> {
+            put(LocalDateTime(2008, 7, 5, 0, 0, 0, 0), ("2008-07-05 00:00:00" to setOf()))
+            put(LocalDateTime(2007, 12, 31, 1, 0, 0, 0), ("2007-12-31 01:00:00" to setOf()))
+            put(LocalDateTime(999, 12, 31, 23, 0, 0, 0), ("0999-12-31 23:00:00" to setOf()))
+            put(LocalDateTime(-1, 1, 2, 0, 1, 0, 0), ("-0001-01-02 00:01:00" to setOf()))
+            put(LocalDateTime(9999, 12, 31, 12, 30, 0, 0), ("9999-12-31 12:30:00" to setOf()))
+            put(LocalDateTime(-9999, 12, 31, 23, 59, 0, 0), ("-9999-12-31 23:59:00" to setOf()))
+            put(LocalDateTime(10000, 1, 1, 0, 0, 1, 0), ("10000-01-01 00:00:01" to setOf()))
+            put(LocalDateTime(-10000, 1, 1, 0, 0, 59, 0), ("-10000-01-01 00:00:59" to setOf()))
+            put(LocalDateTime(123456, 1, 1, 13, 44, 0, 0), ("123456-01-01 13:44:00" to setOf()))
+            put(LocalDateTime(-123456, 1, 1, 13, 44, 0, 0), ("-123456-01-01 13:44:00" to setOf()))
+        }
+        test(LocalDateTimeFormat.fromFormatString("ld<yyyy'-'mm'-'dd> lt<hh:mm:ss>"), dateTimes)
+        test(LocalDateTimeFormat.build {
```

Comment:
while looking on this test, I was thinking about API shape:
1. why not to just use something like this `LocalDateTimeFormat { ... }` for building format? Similar to how kotlinx.serialization create formats (`Json { ... }`) or building`SerializersModule`. Future stdlib `HexFormat` API also uses the same convention https://github.com/JetBrains/kotlin/blob/d3eef139c7b316aa9733f0272c5f8a5f8f844d30/libraries/stdlib/test/text/BytesHexFormatTest.kt#L28
2. What is the reason to add `append` to all methods? I mean, if looking from `formatter` perspective it looks logical (because when formatting, we append elements, so it's ok), but if looking from `parser` perspective it looks odd. IMO, just dropping `append` word looks good from first sight, because then DSL looks like you are just describing the schema for formatting.
So if to apply my comments, creating format for this test will look:
```kotlin
LocalDateTimeFormat {
  year(4)
  literal('-')
  monthNumber(2)
  literal('-')
  dayOfMonth(2)
  literal(' ')
  hour(2)
  literal(':')
  minute(2)
  literal(':')
  second(2)
}
```

If Im missing something, I would be interested to understand :)

Reply:
As is said in the opening post,
> Most importantly: the naming and the shape of user-visible API needs to be discussed.

So, you may well be right, and I can't give an authoritative answer, just explain the reasoning used in the draft. So far, the naming is this way because we'll want to extend the API with localized formats: https://github.com/Kotlin/kotlinx-datetime/discussions/253

> why not to just use something like this `LocalDateTimeFormat { ... }` for building format

I think this format implies that every `LocalDateTimeFormat` can be constructed with this builder, but this may change in the future. I think it doesn't make sense to mix machine communication formats with human-readable formats in a single constructor, given how disparate the use cases for them are, but it also probably would be excessive to have an entirely separate `LocalDateTimeLocalizedFormat` class. Two distinct functions, `build` and `buildLocalized`, seem disparate enough.

Also, we don't have API like `String { +"x"; +"y" }`. Instead, it's `buildString { append("x"); append("y") }`.

> What is the reason to add `append` to all methods?

It's `append`ing to the format. First, the format is "four-digit year," then, it's "four-digit year, followed by a `-`," and so on. It's called `append` to signify that the order is important. When showing a message to the user, you want to reorder the format components, but when parsing logs, you very much don't. If it's `appendYear(4); appendLiteral('-'); appendMonth(2)`, it's clear that it won't be reordered to `appendMonth(2); appendLiteral('-'); appendYear(4)`. If it was just `year`, `literal`, and `month`, I wouldn't bet one way or the other.